### PR TITLE
PGN Viewer to view Games/Replays

### PR DESCRIPTION
**Pull request by rmadhwal**
Hello, thanks again for the great codebase. I took a think about what features I would enjoy, and a replay viewer was my first idea. The remarkable is a device that I use largely to disconnect from the rest of the world, and viewing/studying famous games seems like the most ideal use of the device to me.

So I spent the weekend writing this quick hack. It has worked fine in my limited testing, but I am aware of some limitations that the current implementation would have:

- Only the first ~660 characters of a comment is displayed since otherwise they'd clip into the screen
- Currently the parser that I decided to use (and patch to make it usable enough for this use case) doesn't seem to function well with mulitple arguments or arguments inside paranthesis, so I wrote a dirty regex to get rid of those. Hence, paranthesis comments will be missing from the replay (square brackets comments are fine)
- I have no indication for if a user manually moves and ends up in a checkmate/stalemate, they will simply not be able to play any more moves
- Non standard start states are not well supported (too lazy to implement what would be a fairly niche use case)

Outside of this, since I hacked this over a weekend there may be some bugs that I did not come across in my testing (but I am a good tester generally :) ) I am also aware that I could do a lot more code reuse between `replay_scene` and `game_scene,` in fact I tried to reuse `game_scene` but after some time I felt like I was rewriting it. I am not an experienced rust user as I mentioned earlier and relied largely on my C/C++/Kotlin-Java options knowledge to get me through this (my first) rust change. If you have time, would appreciate comments on what I can improve!

### CODE REVIEW

File: src/bin/chessmarkable/main.rs
Comment (rmadhwal) [inline src/bin/chessmarkable/main.rs:32]: Sorry about this must have been intellij auto-formatting

File: src/bin/chessmarkable/canvas.rs
Comment (rmadhwal) [inline src/bin/chessmarkable/canvas.rs:88]: If performance is a big issue, could optimise this with a binary search, but I was lazy to implement it, do you know if there are any out of the box solutions in Rust ?

**Comment by LinusCDE** [comment]
This looks really awesome! The code looks really promising!

I'll try to test and take a closer look on it tomorrow. I hope it doesn't take multiple days for me to get to do it, but I cant guarantee it.

File: src/bin/chessmarkable/canvas.rs
Comment (LinusCDE) [reply src/bin/chessmarkable/canvas.rs:88]: I don't think that this should be a big concern. The game is already not that fast, since I do compositing images over the framebuffer on a pixel by pixel basis (didn't find a better solution for the rM).

So doing char by char, should work. Didn't take a too close look, but maybe "dry running" each word/character could work to determine whether a potential part would go out of bounds. That could be used to split the text up properly.

I think, I used the "text dry run" method of libremarkable to center text in "buttons".

Comment (LinusCDE) [reply src/bin/chessmarkable/canvas.rs:88]: Not sure myself, whether rust has something for binary search. Probably some sort of sorting libary.

File: src/bin/chessmarkable/main.rs
Comment (LinusCDE) [reply src/bin/chessmarkable/main.rs:32]: I don't think that this is a problem. I did not use rustfmt for this project since I wanted some longer lines for it to be more readable. But I should really have just supplied an according rustfmt.toml file (which IntelliJ would probably use as well) to properly fix this.

**Comment by rmadhwal** [comment]
> This looks really awesome! The code looks really promising!
> 
> I'll try to test and take a closer look on it tomorrow. I hope it doesn't take multiple days for me to get to do it, but I cant guarantee it.

Sure no problem! Would be nice to have someone test it on an RM1 (since I only have an RM2) too.

File: src/bin/chessmarkable/canvas.rs
Comment (rmadhwal) [reply src/bin/chessmarkable/canvas.rs:88]: Yep, I think this may not be a bottleneck, so probably isn't a performance concern. Just thought I'd bring it up in case you want to optimize some parts of the code later.

File: src/bin/chessmarkable/main.rs
Comment (rmadhwal) [reply src/bin/chessmarkable/main.rs:32]: Ah I see, some people mind other people messing with their formatting in their commits so I usually try to undo any automatic changes intellij made during my code review. But if you don't mind this I'll let it be

File: src/bin/chessmarkable/scene/replay_scene.rs
Comment (LinusCDE) [inline src/bin/chessmarkable/scene/replay_scene.rs:363]:
```suggestion
    fn on_input(&mut self, event: InputEvent) {
        match event {
            InputEvent::GPIO { event } => {
                match event {
                    gpio::GPIOEvent::Press { button } => {
                        match button {
                            gpio::PhysicalButton::RIGHT => {
                                // Probably make separate function for it
                                let replay_response = self.replay.play_replay_move();
                                self.update_board(&replay_response.fen);
                                self.clear_state_post_move();
                                self.move_comment = replay_response.comment;
                            },
                            gpio::PhysicalButton::LEFT => {
                                // Probably make separate function for it
                                let replay_response = self.replay.undo_move();
                                self.update_board(&*replay_response.fen);
                                self.clear_state_post_move();
                                self.move_comment = replay_response.comment;
                            },
                            _ => {}
                        }
                    }
                    _ => {}
                }
            },
```

This allows the buttons of the rM 1 to be used like the arrow buttons.

File: src/bin/chessmarkable/pgns.rs
Comment (LinusCDE) [inline src/bin/chessmarkable/pgns.rs:56]: Probably mistyped `pgn` as `png` here.

**Review by LinusCDE — APPROVED** [review]
Hi.

I took a look at the actual application and really like it! I'm personally not really into chess myself ironically and didn't even know about PGN before. The editor is really nice.

I really appreciate the amount of effort that went into this. Many of the rough parts are probably also due to my poor architecture used for this app and I don't consider them a problem for the merge. I probably need to smooth things out eventually if I find the time for it.

Some minor things I noticed:
 - An empty pgn select screen should maybe tell the user where the to put the pgn files (e.g. read out the `CLI_OPTS.pgn_location` for the user since many will just use the gui and probably never know of the cli interface).
 - The above line/comment should probably get removed on going to a prev/next move.
 - Is is more idiomatic to use `replay_response.fen.as_str()` or just `&replay_response.fen` instead of `*&replay_response.fen`
 - `last_move_from` and `last_move_to` in `replace_scene.rs` are not used. Would probably nice to have those visuals as well, but may require a more complex `ReplaceResponse`.

I didn't get to test the game comments feature. I took some random pgns from [PGNMentor](https://www.pgnmentor.com/files.html) and tried some random games.

It works well on the rM 1 as well. I added a suggestion below that assigns the same features to the hardware pageing buttons as the software ones. I'm unsure whether the software ones should remain for the rM 1 or not. I quite like the visual and depending on how people hold the device, they might prefer the software ones over the hardware buttons anyway. Probably just keep them.

20210709_030130_lq

All in all, I quite like the feature already and think it should get merged. Feel free to push any improvements afterwards. (I'll probably have more time in a month or two when my next semester begins.)

:+1:

**Comment by LinusCDE** [comment]
Feel free to merge if you want or keep this open for additional changes. I also don't have any issue on merging later changes.

I leave this up to you. 🙂

**Comment by rmadhwal** [comment]
Thanks for your detailed review and suggestions! Really appreciate it. I agree with all your suggestions and will test and commit them a bit later today.

I completely forgot that the RM1 has buttons and your addition is much appreciated. I think a gesture to get rid of the software buttons for the RM1 could indeed be cool, but since I don't own the device I don't wanna mess around with a change that I can't readily test :) So I'll just slot your changes in for that for now.

**Comment by rmadhwal** [comment]
Ah actually, on your suggestion for the comment to get removed on next move, I disagree a bit since I chose to leave the comment up there on purpose. Sometimes comments can be instructional on the state of the game (rather than just to the move itself), and I prefer text to white space :) Though maybe we can think of a way to make the comment go away in the future (or perhaps refresh can do that)

**Comment by rmadhwal** [comment]
Okay, I think I've covered all your suggestions except the one which removes comments. I think I will revisit that later, need more thought on what the most intuitive behaviour will be. If you think the code is reasonable, would appreciate if you could merge it in (says only those with write access can merge the PR)

**Comment by LinusCDE** [comment]
> Ah actually, on your suggestion for the comment to get removed on next move, I disagree a bit since I chose to leave the comment up there on purpose.

No problem. Was just a suggestion anyway.

I'll merge it. I'll probably make a release out of it the coming days. Or do want me to wait on that? I'm fine with either.

**Comment by rmadhwal** [comment]
Thanks! I'm good with making a release out of this. I'll probably take a stab at the multiplayer/online feature we discussed when I find time next (unless someone discovers a breaking bug) before coming back to improvements for this. Now that I'm a bit more comfortable with the codebase and rust, that seems a bit less daunting :)

**Pull request merged**
